### ISSUE TITLE
Switch to Chumsky for parsing

### ISSUE THREAD

**Issue by hardbyte** [cel-rust/cel-rust#14]
Benefits of using [chumsky](https://github.com/zesterer/chumsky) for parsing:
- Easier to read and modify than LALRPOP grammar
- Much better error reporting and syntax assistance

High level plan:
- [x] Migrate the parser in [hardbyte/common-expression-language](https://github.com/hardbyte/common-expression-language/blob/main/parser/src/parser.rs) to use the same AST as this repository.
- Implement parser for anything missing in https://github.com/hardbyte/common-expression-language
  - [x] recursive members
  - [x] `&&`, `||`, `in` and ternary operations
  - [x] Field Inits - https://github.com/hardbyte/common-expression-language/pull/8
  - [x] byte strings - https://github.com/hardbyte/common-expression-language/pull/7
  - [x] primitive type functions. e.g. `'1'.double()` or `10.double()`
- [ ] Resolve any remaining differences between the existing LALRPOP parser and chumsky one. 
  - [x] Switch this repo's integer implementations to 64bit. https://github.com/clarkmcc/cel-rust/pull/16
  - [ ] Remove the `DoubleMinus` and `DoubleNot`. Should be e.g. `Unary(Negative, Unary(Negative, member))`. https://github.com/clarkmcc/cel-rust/pull/17
  - [ ] others ?
- [x] a small refactor PR in this repository to hide larlpop internals and expose a parser function and parser errors. https://github.com/clarkmcc/cel-rust/pull/12
- [x] more parser tests to increase confidence that a replaced parser doesn't break anything. https://github.com/clarkmcc/cel-rust/pull/13
- [ ] PR to add the chumsky parser as an option perhaps not exposed by default or an optional opt-in?
- [ ] Improvements to error reporting from parser - consider https://github.com/zesterer/ariadne
- [ ] Compare and decide whether to make chumsky parser the default/only parser

Do you want to keep both parsers? If so, how should the API work to pick between them? Assume it wouldn't be too tricky to add unsigned ints and un-escaped strings to the current lalrpop version?

**Comment by clarkmcc** [comment]
> Do you want to keep both parsers?

I'd like to benchmark both and then decide from there. If we need to support both we can cross that bridge. I like the idea of only supporting one parser if possible so it might be worth a performance audit of the chumsky parser if it's significantly different then LALRPOP.

**Comment by hardbyte** [comment]
Yeah that sounds like a good idea. If it isn't too hard to process `textproto` files from rust we could try benchmarking/testing with the [test data](https://github.com/google/cel-spec/tree/master/tests/simple/testdata).

**Comment by clarkmcc** [comment]
Good idea. I think we can use [this function](https://docs.rs/protobuf/latest/protobuf/text_format/fn.parse_from_str.html) to parse those files. I'll need to do some digging to get the full schema of those types, but that shouldn't be difficult to find.

**Comment by hardbyte** [comment]
I'll make another PR to remove the DoubleMinus and DoubleNot, and then I'm down to 2 failing tests with the chumsky parser. 🤞🏼

**Comment by inikolaev** [comment]
Somewhat related to testing topic mentioned here I tried to reuse Gerkin files from [here](https://github.com/cloud-custodian/cel-python/tree/master/features) to run some conformance tests and it works pretty well using the [cucumber-rs](https://github.com/cucumber-rs/cucumber) crate.

I had to make some changes expected output in the feature files, but it more or less works. 

There are some issues with recognising hex and unsigned integers, so I had to adjust the grammar to be able to parse them correctly:

```diff
-    r"-?[0-9]+" => Atom::Int(<>.parse().unwrap()),
-    r"-?0[xX]([0-9a-fA-F]+)" => Atom::Int(i64::from_str_radix(<>, 16).unwrap()),
-    r"-?[0-9]+ [uU]" => Atom::UInt(<>.parse().unwrap()),
-    r"-?0[xX]([0-9a-fA-F]+) [uU]" => Atom::UInt(u64::from_str_radix(<>, 16).unwrap()),
+    r"[-+]?[0-9]+" => Atom::Int(<>.parse().unwrap()),
+    r"([-+])?0[xX]([0-9a-fA-F]+)" => {
+        // We cannot extract capture groups, see https://github.com/lalrpop/lalrpop/issues/575
+        let re = Regex::new(r"([-+])?0[xX]([0-9a-fA-F]+)").unwrap();
+        let captures = re.captures(<>).unwrap();
+
+        let sign = match captures.get(1) {
+            Some(v) => v.as_str(),
+            _ => "",
+        };
+
+        Atom::Int(
+            i64::from_str_radix(
+                format!("{}{}", sign, captures.get(2).unwrap().as_str()).as_str(), 16
+            ).unwrap()
+        )
+    },
+    r"([0-9]+)[uU]" => {
+        // We cannot extract capture groups, see https://github.com/lalrpop/lalrpop/issues/575
+        let re = Regex::new(r"([0-9]+)[uU]").unwrap();
+        let captures = re.captures(<>).unwrap();
+
+        Atom::UInt(captures.get(1).unwrap().as_str().parse().unwrap())
+    },
+    r"0[xX]([0-9a-fA-F]+)[uU]" => {
+        // We cannot extract capture groups, see https://github.com/lalrpop/lalrpop/issues/575
+        let re = Regex::new(r"0[xX]([0-9a-fA-F]+)[uU]").unwrap();
+        let captures = re.captures(<>).unwrap();
+
+        Atom::UInt(u64::from_str_radix(captures.get(1).unwrap().as_str(), 16).unwrap())
+    },
```

But there are other issues as well related to parsing unicode codes like `\U0001f431`, but maybe there's an existing library which can handle those.

The code for the test itself is pretty simple, but it does not support all features yet:

```rust
use cel_interpreter::{Context, Program};
use cucumber::{given, then, when, World};

// `World` is your shared, likely mutable state.
// Cucumber constructs it via `Default::default()` for each scenario.
#[derive(Debug, Default, World)]
pub struct CelWorld {
    expression: String,
}

#[when(expr = "CEL expression \"{word}\" is evaluated")]
fn expression_evaluated_with_double_quotes(world: &mut CelWorld, expression: String) {
    world.expression = expression;
}

#[when(expr = "CEL expression '{word}\' is evaluated")]
fn expression_evaluated_with_single_quotes(world: &mut CelWorld, expression: String) {
    world.expression = expression;
}

#[then(regex = "value is (.*)")]
fn evaluation_result_is(world: &mut CelWorld, expected_result: String) {
    let program = Program::compile(&world.expression).unwrap();
    let mut context = Context::default();
    let result = program.execute(&context);

    assert_eq!(expected_result, format!("{:?}", result));
}

// This runs before everything else, so you can setup things here.
fn main() {
    // You may choose any executor you like (`tokio`, `async-std`, etc.).
    // You may even have an `async` main, it doesn't matter. The point is that
    // Cucumber is composable. :)
    futures::executor::block_on(CelWorld::run("tests/features"));
}
```

and the output looks like this

```
Feature: basic
  Scenario: self_eval_int_zero
   ✔  When CEL expression "0" is evaluated
   ✔  Then value is Ok(Int(0))
  Scenario: self_eval_uint_zero
   ✔  When CEL expression "0u" is evaluated
   ✔  Then value is Ok(UInt(0))
  Scenario: self_eval_float_zero
   ✔  When CEL expression "0.0" is evaluated
   ✔  Then value is Ok(Float(0.0))
  Scenario: self_eval_float_zerowithexp
   ✔  When CEL expression "0e+0" is evaluated
   ✔  Then value is Ok(Float(0.0))
  Scenario: self_eval_string_empty
   ✔  When CEL expression "''" is evaluated
   ✔  Then value is Ok(String(""))
  Scenario: self_eval_string_empty_quotes
   ✔  When CEL expression '""' is evaluated
   ✔  Then value is Ok(String(""))
  Scenario: self_eval_bytes_empty
   ✔  When CEL expression 'b""' is evaluated
   ✔  Then value is Ok(Bytes([]))
  Scenario: self_eval_bool_false
   ✔  When CEL expression "false" is evaluated
   ✔  Then value is Ok(Bool(false))
  Scenario: self_eval_null
   ✔  When CEL expression "null" is evaluated
   ✔  Then value is Ok(Null)
  Scenario: self_eval_empty_list
   ✔  When CEL expression "[]" is evaluated
   ✔  Then value is Ok(List([]))
  Scenario: self_eval_empty_map
   ✔  When CEL expression "{}" is evaluated
   ✔  Then value is Ok(Map(Map { map: {} }))
  Scenario: self_eval_int_nonzero
   ✔  When CEL expression "42" is evaluated
   ✔  Then value is Ok(Int(42))
  Scenario: self_eval_uint_nonzero
   ✔  When CEL expression "123456789u" is evaluated
   ✔  Then value is Ok(UInt(123456789))
  Scenario: self_eval_int_negative_min
   ✔  When CEL expression "-9223372036854775808" is evaluated
   ✔  Then value is Ok(Int(-9223372036854775808))
  Scenario: self_eval_float_negative_exp
   ✔  When CEL expression "-2.3e+1" is evaluated
   ✔  Then value is Ok(Float(-23.0))
  Scenario: self_eval_string_excl
   ✔  When CEL expression '"!"' is evaluated
   ✔  Then value is Ok(String("!"))
  Scenario: self_eval_string_escape
   ✔  When CEL expression "'\''" is evaluated
   ✘  Then value is Ok(String("'"))
      Step failed:
      Defined: tests/features/basic.feature:125:5
      Matched: interpreter/tests/cel.rs:21:1
      Step panicked. Captured output: assertion failed: `(left == right)`
        left: `"Ok(String(\"'\"))"`,
       right: `"Ok(String(\"\\\\'\"))"`
  Scenario: self_eval_bytes_escape
   ✔  When CEL expression "b'ÿ'" is evaluated
   ✔  Then value is Ok(Bytes([195, 191]))
  Scenario: self_eval_bytes_invalid_utf8
   ✔  When CEL expression "b'\000\xff'" is evaluated
   ✘  Then value is BytesType(source=b'\x00\xff')
      Step failed:
      Defined: tests/features/basic.feature:139:5
      Matched: interpreter/tests/cel.rs:21:1
      Step panicked. Captured output: assertion failed: `(left == right)`
        left: `"BytesType(source=b'\\x00\\xff')"`,
       right: `"Ok(Bytes([92, 48, 48, 48, 92, 120, 102, 102]))"`
  Scenario: self_eval_list_singleitem
   ✔  When CEL expression "[-1]" is evaluated
   ✔  Then value is Ok(List([Int(-1)]))
  Scenario: self_eval_map_singleitem
   ✔  When CEL expression '{"k":"v"}' is evaluated
   ✔  Then value is Ok(Map(Map { map: {String("k"): String("v")} }))
  Scenario: self_eval_bool_true
   ✔  When CEL expression "true" is evaluated
   ✔  Then value is Ok(Bool(true))
  Scenario: self_eval_int_hex
   ✔  When CEL expression "0x55555555" is evaluated
   ✔  Then value is Ok(Int(1431655765))
  Scenario: self_eval_int_hex_negative
   ✔  When CEL expression "-0x55555555" is evaluated
   ✔  Then value is Ok(Int(-1431655765))
  Scenario: self_eval_uint_hex
   ✔  When CEL expression "0x55555555u" is evaluated
   ✔  Then value is Ok(UInt(1431655765))
  Scenario: self_eval_unicode_escape_four
   ✔  When CEL expression '"\u270c"' is evaluated
   ✘  Then value is Ok(String("✌"))
      Step failed:
      Defined: tests/features/basic.feature:188:5
      Matched: interpreter/tests/cel.rs:21:1
      Step panicked. Captured output: assertion failed: `(left == right)`
        left: `"Ok(String(\"✌\"))"`,
       right: `"Ok(String(\"\\\\u270c\"))"`
  Scenario: self_eval_unicode_escape_eight
   ✔  When CEL expression '"\U0001f431"' is evaluated
   ✘  Then value is Ok(String("🐱"))
      Step failed:
      Defined: tests/features/basic.feature:195:5
      Matched: interpreter/tests/cel.rs:21:1
      Step panicked. Captured output: assertion failed: `(left == right)`
        left: `"Ok(String(\"🐱\"))"`,
       right: `"Ok(String(\"\\\\U0001f431\"))"`
  Scenario: self_eval_ascii_escape_seq
   ✔  When CEL expression '"\a\b\f\n\r\t\v\"\'\\"' is evaluated
   ✘  Then value is String(\x07\x08\x0c\n\r\t\x0b"\'\\)
      Step failed:
      Defined: tests/features/basic.feature:202:5
      Matched: interpreter/tests/cel.rs:21:1
      Step panicked. Captured output: assertion failed: `(left == right)`
        left: `"String(\\x07\\x08\\x0c\\n\\r\\t\\x0b\"\\'\\\\)"`,
       right: `"Ok(String(\"\\\\a\\\\b\\\\f\\\\n\\\\r\\\\t\\\\v\\\\\\\"\\\\'\\\\\\\\\"))"`
  Scenario: self_eval_bound_lookup
   ?  Given type_env parameter "x" is TypeType(value='INT64')
      Step skipped: tests/features/basic.feature:211:4
  Scenario: self_eval_unbound_lookup
   ✔  When CEL expression "x" is evaluated
   ?  Then eval_error is "undeclared reference to 'x' (in container '')"
      Step skipped: tests/features/basic.feature:225:5
  Scenario: unbound_is_runtime_error
   ?  When CEL expression "x || true" is evaluated
      Step skipped: tests/features/basic.feature:230:5
  Scenario: binop
   ?  When CEL expression "1 + 1" is evaluated
      Step skipped: tests/features/basic.feature:240:5
  Scenario: unbound
   ✔  When CEL expression "f_unknown(17)" is evaluated
   ?  Then eval_error is 'unbound function'
      Step skipped: tests/features/basic.feature:249:5
  Scenario: unbound_is_runtime_error
   ?  When CEL expression "f_unknown(17) || true" is evaluated
      Step skipped: tests/features/basic.feature:254:5
  Scenario: false
   ?  Given type_env parameter "false" is TypeType(value='BOOL')
      Step skipped: tests/features/basic.feature:265:4
  Scenario: true
   ?  Given type_env parameter "true" is TypeType(value='BOOL')
      Step skipped: tests/features/basic.feature:278:4
  Scenario: null
   ?  Given type_env parameter "null" is TypeType(value='BOOL')
      Step skipped: tests/features/basic.feature:291:4
[Summary]
1 feature
37 scenarios (23 passed, 9 skipped, 5 failed)
67 steps (53 passed, 9 skipped, 5 failed)
```

**Comment by clarkmcc** [comment]
@inikolaev this is fantastic! Definitely makes sense to incorporate into the project. I took a look at the proto-based tests but read an announcement from the maintainer about deprecating the protobuf dependency so I stopped pursuing that. This looks great however.

**Comment by clarkmcc** [comment]
Irrelevant due to using ANTLR now

**Comment by polarathene** [comment]
> Irrelevant due to using ANTLR now

Linking for context:
- https://github.com/cel-rust/cel-rust/pull/136#pullrequestreview-2918797166 (_switching to ANTLR better aligns with other CEL implementations_)
- https://github.com/cel-rust/cel-rust/issues/140
- https://github.com/antlr/antlr4
- https://github.com/google/cel-spec/issues/63#issuecomment-823254728
- https://pkg.go.dev/github.com/google/cel-go@v0.26.1/parser/gen (CEL uses ANTLR)

---

> Benefits of using [chumsky](https://github.com/zesterer/chumsky) for parsing:
> 
>     * Easier to read and modify than LALRPOP grammar
> 
>     * Much better error reporting and syntax assistance

Compared to the original cited incentive for using Chumsky, I don't know if those were addresed by the switch to ANTLR? Might be useful context for anyone curious.

@alexsnaps didn't provide much context on the matter with referenced issues / PR (or original PR that #140 referenced. Perhaps there was some discussion elsewhere before that PR landed, but I'm not sure where if it was on Github somewhere 😅 (_I am aware of pursuing support for [a Rust target into the upstream ANTLR project](https://github.com/antlr/antlr4/issues/1839#issuecomment-2919227018)_)

**Comment by alexsnaps** [comment]
@polarathene I guess when you say "I don't know if those were addressed by the switch to ANTLR" you are referring to these two points, right?

 1. Easier to read and modify than LALRPOP grammar
 1. Much better error reporting and syntax assistance

So let me start with answering these:

 1. There is no need to modify the grammar, the grammar we use here is from [the reference implementation](https://github.com/google/cel-go/blob/master/parser/gen/CEL.g4), so does the [Java](https://github.com/google/cel-java/blob/main/parser/src/main/java/dev/cel/parser/gen/CEL.g4) & the [C++ one](https://github.com/google/cel-java/blob/main/parser/src/main/java/dev/cel/parser/gen/CEL.g4).
 What we want here is a compliant _and_ compatible AST produced. More on that below.
 1. I guess that can be up for debate... but I understand this as "parse errors reported to the user", where I think we've come [a long way](https://github.com/cel-rust/cel-rust/blob/master/cel/src/parser/parser.rs#L1946-L1956), don't you think? The [`ParseError`s](https://github.com/cel-rust/cel-rust/blob/master/cel/src/parser/parser.rs#L74-L80) reported each contain pretty much one could desire, no? What would you add? Obviously if there error reporting can be made better, we should try to do so!

On the AST tho, because that's an important goal, CEL isn't designed for fastest parsing necessarily. One goal is to have [persistable AST ready for fast execution](https://cel.dev/overview/cel-overview#phases_of_expression_processing). There is a protobuf representation, meant to be portable across interpreter implementations, that exist (but that we don't yet support). Also, having a comparable AST makes implementing the interpreter in a spec compliant manner a whole lot easier. I've spent some time in the past, including changing the grammar, to get to a place where we could "mimic the other interpreter's behavior as closely as possible", but that's quite a challenge. 

Now, not to say that the parsing shouldn't be as fast as possible. But the resulting AST should pass the [test cases](https://github.com/cel-rust/cel-rust/blob/master/cel/src/parser/parser.rs#L1257), which actually come from the golang impl., just as much. 

tl;dr if there is a clear advantage to using a different parser, we should consider the option to: initially have it live side by side with the antlr4 one while getting it up to par; then possibly consider dropping the "less good one". But know that the grammar evolves and requires updates when the upstream do change it. Today, this simply requires copying the `.g4` over and regenerating the parser with antlr. 

I hope this clarifies somewhat and explains some of the rational, while addressing open questions on your end.

**Comment by polarathene** [comment]
Thanks for the clarification 😁 Makes sense with the switch to ANTLR 👍 

I don't have anything else to add other than some of your links might be better as permalinks, especially when referencing lines of source code, since that's fairly likely to shift over time, should any reader in future be interested in the referenced lines.
